node/toolrunner.ts

```diff
@@ -580,200 +580,234 @@ export class ToolRunner extends events.EventEmitter {
-                    this._debug(toolPath + ' might have exited due to errors prematurely. Verify the arguments passed are valid.');
+                    catch (err) {
+                        fileStreamError = true;
+                        this._debug(`Failed to pipe the output of '${this.toolPath}' to the file '${this.pipeOutputToFile}'. Error: ${err}`);
+                    }
+                }
+
+                if (!stdinError) {
+                    try {
+                        cp2.stdin.write(data);
+                    }
+                    catch (err) {
+                        this._debug(`Failed to pipe the output of '${this.toolPath}' to '${this.pipeOutputToTool.toolPath}'. The tool '${this.pipeOutputToTool.toolPath}' might have exited due to errors. Verify the arguments passed are valid. Error: ${err}`);
+                    }
                 }
             });
-            cpFirst.stderr.on('data', (data: Buffer) => {
+
+            cp1.stderr.on('data', (data: Buffer) => {
+                state.processStderr = true;
+
                 if (fileStream) {
                     fileStream.write(data);
                 }
-                successFirst = !options.failOnStdErr;
+
                 if (!options.silent) {
                     var s = options.failOnStdErr ? options.errStream : options.outStream;
                     s.write(data);
                 }
             });
-            cpFirst.on('error', (err) => {
-                waitingEvents--; //first process is complete with errors
+
+            cp1.on('error', (err) => {
+                state.processError = err.message;
+                state.processExited = true;
+                state.processClosed = true;
+
                 if (fileStream) {
                     fileStream.end();
                 }
-                cp.stdin.end();
-                error = new Error(toolPathFirst + ' failed. ' + err.message);
-                if(waitingEvents == 0) {
-                    defer.reject(error);
-                }
+
+                cp2.stdin.end();
+                state.CheckComplete();
             });
-            cpFirst.on('close', (code, signal) => {
-                waitingEvents--; //first process is complete
-                if (code != 0 && !options.ignoreReturnCode) {
-                    successFirst = false;
-                    returnCodeFirst = code;
-                    returnCode = returnCodeFirst;
-                }
-                this._debug('success of first tool:' + successFirst);
+
+            cp1.on('exit', (code, signal) => {
+                state.processExitCode = code;
+                state.processExited = true;
+                this._debug(`Exit code ${code} received from tool '${this.toolPath}'`);
+                state.CheckComplete()
+            });
+
+            cp1.on('close', (code, signal) => {
+                state.processExitCode = code;
+                state.processClosed = true;
+
```

Comment:
do we also want to set `state.processExited = true;` in case we miss the exit event.